### ISSUE TITLE
Window decorations on Gnome/Wayland

### ISSUE THREAD

**Issue by last-partizan** [rust-windowing/winit#1967]
Hello.

Recently neovide started to use wayland for rendering, and it uses those ugly decorations mentioned in #306 

I see winit switched to using client-toolkit for rendering, and it makes no attempt to give any visual quality, but acts only as fallback.

- https://github.com/Smithay/client-toolkit/issues/162#issuecomment-798955486

So, where is ther right place to fix those fallback decorations?

I created [discussion at discourse.gnome.org](https://discourse.gnome.org/t/proper-way-to-render-window-decorations-on-wayland-for-non-gtk-apps/6734) and someone suggested that winit should use [libdecor](https://gitlab.gnome.org/jadahl/libdecoration) to implement gtk decorations.

Could this be implemented?

By the way, is there some way to ask winit not to use wayland, even if it's available? That would fix issue with decorations as workaround, until it's resolved.

**Comment by last-partizan** [comment]
https://github.com/Kethku/neovide/issues/728

Linking this issue.

**Comment by alexheretic** [comment]
> By the way, is there some way to ask winit not to use wayland, even if it's available? That would fix issue with decorations as workaround, until it's resolved.

You can force xwayland usage with env var `WINIT_UNIX_BACKEND=x11`, see [docs](https://docs.rs/winit/0.25.0/winit/event_loop/struct.EventLoop.html#method.new).

**Comment by maroider** [comment]
Using `libdecor` as a fallback sounds interesting, but it may be better to have such an integration implemented upstream in SCTK.

cc @vberger

**Comment by vberger** [comment]
I purposefully will not support proper pretty decorations directly in SCTK, because that would require SCTK itself to depend on a whole font rendering engine to just to write text in the titlebar, and decorations look and feel is a complex matter on which everyone has its own opinion.

This is why SCTK provides [a trait](https://docs.rs/smithay-client-toolkit/0.14.0/smithay_client_toolkit/window/trait.Frame.html) that can be implemented to create an draw decorations as desired. This trait could be implemented in winit, or as a standalone crate. I just don't plan to have any font rendering in SCTK, that is way out of its scope.

Implementing the use of `libdecor` when available would, however, be in scope for SCTK.

**Comment by vberger** [comment]
Looking at the `neovide` issue, if people are complaining that winit's decorations are not coherent with the rest of GTK decorations, there are only two ways out of that: either neovide could witch to GTK from winit, of SCTK can implement `libdecor` support. However I have no idea how much `libdecor` is actually packaged in distributions for people to use.

**Comment by maroider** [comment]
Ubuntu doesn't seem to distribute it, and neither does Arch. There is an AUR package ([`libdecoration-git`](https://aur.archlinux.org/packages/libdecoration-git/)), but that's about it.

**Comment by floatingrain** [comment]
So how should we deal with this problem now?

**Comment by kchibisov** [comment]
I don't see an issue here in general, since it's a gnome folks decision to make everyone draw client side decorations and last time I've asked them they replied in a way that I should draw not what GTK does, but what I think suits the app most for its decorations, which makes sense. In winit we have decorations as some sort of fallback, meaning that it doesn't really matter how they look in the end, they should just be, and that's it.

However not providing a way to change them isn't something I want to end up without, however providing 'hooks' for surfraces, libdecor support, whatever isn't really a priority for me, since I don't really care about wayland servers don't provide server side decorations(you won't get consistency without them anyway).

To sum up, I'd go with `FallbackFrame` as for sctk update, however I'd like to see a way to handle wayland decorations in winit nicely, which means without blowing tons of wayland internals on the end user, and that part could be hard.

**Comment by floatingrain** [comment]
However, the default decoration of winit cannot produce shadow effects around windows. If I want to make hierarchical shadows between windows, what should I do?

**Comment by JMS55** [comment]
As an update to this, @cmeissl is working on a [libdecor-rs](https://github.com/cmeissl/libdecor-rs). Once that's done, I imagine we could integrate it with SCTK/Winit.

However, it looks like they ran into some [safety issues](https://gitlab.gnome.org/jadahl/libdecor/-/issues/30) with libdecor that need to get resolved first.

**Comment by satmandu** [comment]
Looks like debian & ubuntu also package libdecor: 
https://salsa.debian.org/sdl-team/libdecor-0
https://packages.ubuntu.com/ca/source/jammy/libdecor-0
